File: files/000-099/035/CH-035.R
library(tidyverse)
library(readxl)

input = read_excel("files/CH-035 Up and Down Grades.xlsx", range = "B2:D37")
test  = read_excel("files/CH-035 Up and Down Grades.xlsx", range = "H2:K6")

result = input %>%
  mutate(month = floor_date(Date, "month"),
         Grade = factor(Grade, levels = c("A+","A", "B", "C"), ordered = TRUE)) %>%
  summarise(last_grade = last(Grade), .by = c(`Agent-id`, month)) %>%
  arrange(`Agent-id`, month) %>%
  mutate(prev_grade = lag(last_grade), .by = `Agent-id`) %>%
  filter(!is.na(prev_grade)) %>%
  mutate(transition = case_when(
    prev_grade > last_grade ~ "Upgrade",
    prev_grade < last_grade ~ "Down-grade",
    TRUE ~ "No change"
  )) %>%
  mutate(month = as.numeric(factor(month))) %>%
  summarise(transitions = n(), .by = c(transition, month)) %>%
  pivot_wider(names_from = transition, values_from = transitions, values_fill = 0) %>%
  arrange(month) %>%
  select(Month = month, Upgrade, `No change`, `Down-grade`)

print(result)
print(test)

  

File: files/000-099/035/CH-035.py
import pandas as pd

input = pd.read_excel("CH-035 Up and Down Grades.xlsx", sheet_name="Sheet1", usecols="B:D", skiprows=1)
test = pd.read_excel("CH-035 Up and Down Grades.xlsx", sheet_name="Sheet1", usecols="H:K", skiprows=1, nrows=4)

result = input.assign(
    Month=pd.to_datetime(input['Date']).dt.to_period('M'),
    Grade=pd.Categorical(input['Grade'], categories=["A+", "A", "B", "C"], ordered=True)
).groupby(['Agent-id', 'Month']).agg(last_grade=('Grade', 'last')).reset_index()

result['prev_grade'] = result.groupby('Agent-id')['last_grade'].shift()
result = result.dropna(subset=['prev_grade'])

grades = {"A+": 4, "A": 3, "B": 2, "C": 1}
result['prev_grade'] = result['prev_grade'].map(grades)
result['last_grade'] = result['last_grade'].map(grades)
result['transition'] = ['Upgrade' if x > y else 'Down-grade' if x < y else 'No Change' for x, y in zip(result['last_grade'], result['prev_grade'])]

result['Month'] = result['Month'].dt.month
result = result.groupby(['transition', 'Month']).size().reset_index(name='transitions')
result = result.pivot(index='Month', columns='transition', values='transitions').fillna(0).reset_index().sort_values('Month')
result = result[['Month', 'Upgrade', 'No Change', 'Down-grade']].astype({"Upgrade": "int64", "No Change": "int64", "Down-grade": "int64"})

print(result == test)  # Discrepancy in month 3
